doc/api/async_hooks.md

```diff
@@ -0,0 +1,677 @@
+execution timing. Not causality (which is covered by `triggerId()`). For
+example:
+
+```js
+const server = net.createServer(function onConnection(conn) {
+  // Returns the id of the server, not of the new connection. Because the
+  // on connection callback runs in the execution scope of the server's
+  // MakeCallback().
+  async_hooks.currentId();
+
+}).listen(port, function onListening() {
+  // Returns the id of a TickObject (i.e. process.nextTick()) because all
+  // callbacks passed to .listen() are wrapped in a nextTick().
+  async_hooks.currentId();
+});
+```
+
+#### `async_hooks.triggerId()`
+
+* Returns {number} the id of the resource responsible for calling the callback
+  that is currently being executed.
+
+For example:
+
+```js
+const server = net.createServer(conn => {
+  // Though the resource that caused (or triggered) this callback to
+  // be called was that of the new connection. Thus the return value
+  // of triggerId() is the id of "conn".
+  async_hooks.triggerId();
+
+}).listen(port, () => {
+  // Even though all callbacks passed to .listen() are wrapped in a nextTick()
+  // the callback itself exists because the call to the server's .listen()
+  // was made. So the return value would be the id of the server.
+  async_hooks.triggerId();
+});
+```
+
+## Embedder API
+
+Library developers that handle their own I/O will need to hook into the
+AsyncWrap API so that all the appropriate callbacks are called. To accommodate
+this a JavaScript API is provided.
+
+### `class AsyncEvent()`
+
+The class `AsyncEvent` was designed to be extended by the embedder's async
+resources. Using this users can easily trigger the lifetime events of their
+own resources.
+
+The `init()` hook will trigger when an `AsyncEvent` is instantiated.
+ 
+It is important that before/after calls are unwound
+in the same order they are called. Otherwise an unrecoverable exception
+will be made.
+
+```js
+// AsyncEvent() is meant to be extended. Instantiating a new AsyncEvent() also
+// triggers init(). If triggerId is omitted then currentId() is used.
+const asyncEvent = new AsyncEvent(type[, triggerId]);
+
+// Call before() hooks.
+asyncEvent.emitBefore();
+
+// Call after() hooks.
+asyncEvent.emitAfter();
+
+// Call destroy() hooks.
+asyncEvent.emitDestroy();
+
+// Return the unique id assigned to the AsyncEvent instance.
+asyncEvent.asyncId();
+
+// Return the trigger id for the AsyncEvent instance.
+asyncEvent.triggerId();
+```
+
+#### `AsyncEvent(type[, triggerId])`
+
+* arguments
+  * `type` {string} the type of ascycn event
+  * `triggerId` {number} the id of the execution context that created this async
+    event
+* Returns {AsyncEvent} A reference to `asyncHook`.
+
+Example usage:
+
+```js
+class DBQuery extends AsyncEvent {
+  constructor(db) {
+    this.db = db;
+  }
+
+  getInfo(query, callback) {
+    this.db.get(query, (err, data) => {
+      this.emitBefore();
+      callback(err, data)
+      this.emitAfter();
```

Comment:
`callback` is user code, if it throws, `after()` will never be reached, which will break the API contract ("exactly once"), isn't it necessary to use try/finally here? People will cut-n-paste this usage directly into their packages, it _must_ be exactly right! And if it is exactly right, better describe why try/catch-or-finally is not necessary.